### PR TITLE
Add type annotations with mypy

### LINKED ISSUE #2080: add type annotations to Python code
Start with annotating the types of the most common functions (important interfaces like `bundle_model.py`).

In some places, we use tuples to represent worksheet items and worksheet_info is a complex dictionary.  We should consider using `TypedDict` for that.  These are probably places where annotations would help a lot.

### PR DESCRIPTION
**Pull request by yiboliu**
### Reasons for making this change
1. Resolve Exception must be derived from BaseException
2. Resolve Need type annotation for '*'
3. Resolve Cannot find implementation or library stub for module named '*'
4. Add the requirements.dev.txt into installation list
5. Further comments will be added if there are more to commit here

### Related issues
Resolves #2080 

### Checklist

* [x] I've added a screenshot of the changes, if this is a frontend change
* [x] I've added and/or updated tests, if this is a backend change
* [x] I've run the [pre-commit.sh](https://github.com/codalab/codalab-worksheets/blob/master/pre-commit.sh) script
* [x] I've updated [docs](https://github.com/codalab/codalab-worksheets/tree/master/docs), if needed

### CODE REVIEW

File: codalab/worker/rest_client.py [hunk 1]
```diff
@@ -27,7 +28,7 @@ class RestClient(object):
     """
     A dictionary of additional headers to send along with HTTP requests.
     """
-    _extra_headers = {}
+    _extra_headers: Dict[Any, Any] = {}
```
Comment (percyliang) [inline codalab/worker/rest_client.py]: Can the types be more specific?  At least the key is a string, no?

File: requirements-server.txt [hunk 2]
```diff
@@ -8,7 +8,7 @@
 -r requirements.dev.txt
 
 # For worker manager
-boto3==1.9.228
+boto3~=1.14.33
```
Comment (percyliang) [inline requirements-server.txt]: Are these part of this PR?

**Review by percyliang — COMMENTED** [review]
Can you write a PR description saying more about what the scope of this PR is?  Why these particular additions and these files?  It doesn't seem like it's complete, so does this correspond to some conceptual subset of work?

File: requirements-server.txt (hunk 2 above)
Comment (yiboliu) [reply requirements-server.txt]: Yes, it's upgrading a boto3 version. It's suggested by pycharm.

File: codalab/worker/rest_client.py (hunk 1 above)
Comment (yiboliu) [reply codalab/worker/rest_client.py]: sure, I can do it.

File: requirements-server.txt [hunk 3]
```diff
@@ -17,5 +17,5 @@ nose==1.3.7
 # Server-specific
 alembic==1.0.0
 gunicorn==20.0.4
-oauthlib==2.1.0
+oauthlib~=3.1.0
```
Comment (teetone) [inline requirements-server.txt]: I think this PR should just be about mypy related changes. Also,  we tried to upgrade `oauthlib` to `v3.1.0` and it broke our authentication. See this recent commit here: https://github.com/codalab/codalab-worksheets/commit/1f2c0ea6fadc0eaf5e86bc3d5b407c146bf348d4

File: requirements.txt [hunk 4]
```diff
@@ -13,3 +13,5 @@ fusepy==2.0.4
 python-dateutil==2.8.1
 diffimg==0.2.3
 selenium==3.141.0
+
+requests~=2.24.0
```
Comment (teetone) [inline requirements.txt]: Can you explain what this library is for?

File: codalab/lib/bundle_fuse.py [hunk 5]
```diff
@@ -8,6 +8,8 @@
 from collections import OrderedDict
 
 from contextlib import closing
+from typing import Dict, Any
```
Comment (teetone) [inline codalab/lib/bundle_fuse.py]: Is `Any` being used in this file?

File: test_cli.py [hunk 6]
```diff
@@ -386,7 +388,7 @@ class TestModule(object):
     a decorator to register new modules and a class method to run modules by name.
     """
 
-    modules = OrderedDict()
+    modules: Dict[str, Any] = OrderedDict()
```
Comment (teetone) [inline test_cli.py]: We should try to avoid `Any` as much as we can. I think `Any` is the same as not specifying a type. Should it be `TestModule` instead?

File: requirements.txt (hunk 4 above)
Comment (yiboliu) [reply requirements.txt]: There are a few files importing this library, like codalab/rest/interpret.py, codalab/worker/docker_utils.py, etc

File: requirements-server.txt (hunk 3 above)
Comment (yiboliu) [reply requirements-server.txt]: yeah, I can roll it back.

File: codalab/lib/bundle_fuse.py (hunk 5 above)
Comment (yiboliu) [reply codalab/lib/bundle_fuse.py]: I removed it and usages.

File: test_cli.py (hunk 6 above)
Comment (yiboliu) [reply test_cli.py]: Are we using python 3.7 or newer? If not we can't refer to the class within itself. I worked out another way.

Comment (percyliang) [reply test_cli.py]: You can use a string: `"TestModule"`

Comment (yiboliu) [reply test_cli.py]: Yeah, I did it in my latest update.

File: requirements-server.txt (hunk 2 above)
Comment (epicfaace) [inline requirements-server.txt]: Can you remove the `~=` in all dependencies? We want to pin dependencies to exact versions using `==` instead.

Also, are these dependency upgrades required for the mypy checks to work? If not, we should do them in a separate PR (dependabot might have already caught a few of them).

**Review by epicfaace — CHANGES_REQUESTED** [review]

**Review by teetone — APPROVED** [review]

**Review by epicfaace — APPROVED** [review]
